contracts/rollover/CrossCurrencyRollover.sol

```diff
@@ -0,0 +1,428 @@
+            revert CCR_CollateralMismatch(
+                sourceLoanTerms.collateralAddress,
+                sourceLoanTerms.collateralId,
+                newLoanTerms.collateralAddress,
+                newLoanTerms.collateralId
+            );
+        }
+
+        // ------------- New LoanTerms Validation -------------
+        originationHelpers.validateLoanTerms(newLoanTerms);
+    }
+
+    // ========================================= HELPERS ================================================
+    /**
+     * @notice Executes the rollover of a loan from one set of terms to another.
+     *         Handles the settlement of old loan amounts and the necessary funds transfer
+     *         to repay the old loan and initialize the new loan.
+     *
+     * @param oldLoanId                  The ID of the original loan to be rolled over.
+     * @param oldLoanData                The loan data of the original loan.
+     * @param lender                     The address of the new loan lender.
+     * @param newLoanTerms               The terms of the new loan.
+     * @param swapParams                 The parameters for the currency swap.
+     */
+    function _executeRollover(
+        uint256 oldLoanId,
+        LoanLibrary.LoanData memory oldLoanData,
+        address lender,
+        LoanLibrary.LoanTerms memory newLoanTerms,
+        OriginationLibrary.SwapParameters memory swapParams
+    ) internal {
+        // get funds for new loan and swap them
+        (uint256 repayAmount, LoanLibrary.FeeSnapshot memory feeSnapshot) = _processSettlement(
+            msg.sender,
+            lender,
+            swapParams,
+            newLoanTerms,
+            oldLoanData,
+            oldLoanId
+        );
+
+        // repay old loan
+        _repayLoan(msg.sender, IERC20(oldLoanData.terms.payableCurrency), oldLoanId, repayAmount);
+
+        // initialize new loan
+        _initializeRolloverLoan(newLoanTerms, msg.sender, lender, oldLoanId, feeSnapshot);
+    }
+
+    /**
+     * @notice Initializes the new loan.
+     *
+     * @param newTerms                  The terms of the new loan.
+     * @param borrower                  The address of the borrower.
+     * @param lender                    The address of the lender.
+     * @param oldLoanId                 The ID of the original loan.
+     * @param feeSnapshot               Principal and interest fees pertaining to rollover.
+     *
+     * @return newLoanId                The ID of the new loan.
+     */
+    function _initializeRolloverLoan(
+        LoanLibrary.LoanTerms memory newTerms,
+        address borrower,
+        address lender,
+        uint256 oldLoanId,
+        LoanLibrary.FeeSnapshot memory feeSnapshot
+    ) internal nonReentrant returns (uint256 newLoanId) {
+        LoanLibrary.LoanData memory oldLoanData = ILoanCore(loanCore).getLoan(oldLoanId);
```

Comment:
We already get the oldLoanData in `rolloverCrossCurrencyLoan()` can it be passed as input to this function? Will save an external call.

Reply:
It's not even used in this function, so it can just be deleted here.

Reply:
Good catch.